src/webgpu/api/validation/encoding/queries/general.spec.ts

```diff
@@ -5,10 +5,100 @@ TODO:
     - queryIndex {in, out of} range for GPUQuerySet
     - GPUQuerySet {valid, invalid}
         - or {undefined}, for occlusionQuerySet
-    - ?
+    - x = {occlusion, pipeline statistics, timestamp} query
 `;
 
+import { params, poptions } from '../../../../../common/framework/params_builder.js';
 import { makeTestGroup } from '../../../../../common/framework/test_group.js';
+import { kQueryTypes } from '../../../../capability_info.js';
 import { ValidationTest } from '../../validation_test.js';
 
+async function selectDeviceForQueryType(t: ValidationTest, type: GPUQueryType): Promise<void> {
+  const extensions: GPUExtensionName[] = [];
+  if (type === 'pipeline-statistics') {
+    extensions.push('pipeline-statistics-query');
+  } else if (type === 'timestamp') {
+    extensions.push('timestamp-query');
+  }
+
+  await t.selectDeviceOrSkipTestCase({ extensions });
+}
+
+export const enum EncoderType {
+  CommandEncoder = 'CommandEncoder',
+  ComputeEncoder = 'ComputeEncoder',
+  RenderEncoder = 'RenderEncoder',
+}
+
 export const g = makeTestGroup(ValidationTest);
+
+g.test('writeTimestamp')
+  .desc(
+    `
+Tests that write timestamp to all types of query set on all possible encoders:
+- queryIndex {in, out of} range for GPUQuerySet
+- GPUQuerySet {valid, invalid}
+- x= {occlusion, pipeline statistics, timestamp} query
+- x= {compute, render, non-pass} enconder
+  `
+  )
+  .params(
```

Comment:
Please use `.cases` and `.subcases`. In my opinion, `encoderType` and `type` should be cases, while `queryIndex` should be subcases.